### PR TITLE
[proposal] zero-allocation Set

### PR DESCRIPTION
**Pull request by MrAlias**
This proposes to separate the data a `Set` represents from the `Set` itself. Doing so allows for ...

- zero(-ish) allocations to the heap and little computation overhead during set construction
- reduction in size of data when passing `Set`s (i.e. the metric pipeline)

It also means that while `Set`s remain comparable, their equivalents can no longer be tested with `==`. The existing `Distinct` type or the `Equals` method need to be used to test equivalence.

## Design

- Add a registry to globally hold all `Set` data.
  - `Set`s are constructed by sorting and de-duplicating their data then passing that data to the registry for a unique ID.
  - The data in the registry uses reference counting to know when it can be removed from the registry
- Redefine a `Set` to hold pointer to a unique ID of the data (note: this is not the same **reference** as other sets for the same data)
   - When `Set`s are constructed their IDs have a finalizer set. When they become unreachable the reference they hold to the data will be removed from the registry.
- Redefine a `Destinct` to hold the unique ID of the `Set` data the `Destinct` was made from
- Use pools for all set data and IDs to amortize heap allocations
- A zero allocation implementation of the FNV-1a hash algorithm is added to compute unique IDs for the `Set` data

### Zero Allocations

By owning all of the data and controlling its life-cycle, we are able to effectively use pools for any data that needs to be allocated to the head. This means that all allocations can be amortized and effectively creating `Set`s will require zero allocations to the heap.

### Reduced size of `Set`

Defining a `Set` with a single `*uint64` field means the size of the `Set` is now 8 bytes. Contrast this to the prior implementation (an `interface{}` (2 `uintptr`) referencing an array of `N` `KeyValue`) which was `(2 * sizeOf(uintptr)) + (N * sizeOf(KeyValue))` (on a 32-bit system with no data this would be 8 bytes).

This reduction in `Set` data size means that whenever a set is passed as an argument, the sized copied on the stack will be much smaller, but also consistent. This means the `Set` will always be able to be in a [small size stack frame](https://go.dev/src/runtime/stack.go).

#### Equivalence testing

The `Set` is still comparable. However, comparison of `Set`s created with the same `KeyValue`s will evaluate to `false` when compared in a `map` or with `==`. To test equivalence the `Equals` method needs to be used, and a `map` should be defined over the `Distinct` type instead.

This incompatibility does not break the API, but it does alter released behavior. Even though the `Set` is defined with an `Equals` method and the `Distinct` type is explicitly declared to be used as a map key instead, this will likely break user (and our metric pipeline) code. Careful consideration of if this is acceptable needs to be made.

It might be possible to resolve this, but I have not found a way. More investigation might be beneficial.

### Testing

- Moderate test coverage is included. It is not release-ready level of testing.
- The `BenchmarkNewSet` benchmark is added to show the performance improvement of the changes.

```sh
$ go test -run='^$' -bench=BenchmarkNewSet -count=20 > out.txt
$ benchstat out.txt
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/attribute
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
         │   out.txt    │
         │    sec/op    │
NewSet-8   2.327µ ± 30%

         │   out.txt   │
         │    B/op     │
NewSet-8   19.50 ± 18%

         │  out.txt   │
         │ allocs/op  │
NewSet-8   0.000 ± 0%
```

### CODE REVIEW

**Comment by MrAlias** [comment]
An additional branch from this that would show the SDK changes needed to support this would be helpful.

**Comment by pellared** [comment]
> Even though the Set is defined with an Equals method and the Distinct type is explicitly declared to be used as a map key instead, this will likely break user (and our metric pipeline) code.

Do I understand correctly that this will break if one uses a new version of `go.opentelemetry.io/otel` (with new Set implementation) and existing version of `go.opentelemetry.io/otel/sdk/metric` (which is NOT stable yet). As this is not stable I think it could be acceptable. The question is if it would break any stable Go module?

File: attribute/set.go [hunk 1]
```diff
@@ -53,16 +58,44 @@ type (
 	Sortable []KeyValue
 )
 
-var (
-	// keyValueType is used in computeDistinctReflect.
-	keyValueType = reflect.TypeOf(KeyValue{})
+var slicePool = sync.Pool{New: func() any { return new([]KeyValue) }}
```
Comment (tigrannajaryan) [inline attribute/set.go:61]: I am curious how much this pool helps with performance. How much worse it is without this pool and with just allocating slices as needed?

File: attribute/set.go [hunk 2]
```diff
@@ -53,16 +58,44 @@ type (
...
 
-	// emptySet is returned for empty attribute sets.
-	emptySet = &Set{
-		equivalent: Distinct{
-			iface: [0]KeyValue{},
-		},
+func getSlice(length, capacity int) *[]KeyValue {
+	v := slicePool.Get().(*[]KeyValue)
+	if cap(*v) < capacity {
+		*v = make([]KeyValue, length, capacity)
```
Comment (tigrannajaryan) [inline attribute/set.go:66]:
```suggestion
		return make([]KeyValue, length, capacity)
```

Comment (tigrannajaryan) [reply attribute/set.go:66]: Also can return v to the pool since we didn't use it.

File: attribute/set.go [hunk 3]
```diff
@@ -53,16 +58,44 @@ type (
...
+	id := getID()
+	*id = sets.Store(data)
+	// Set the finalizer here, instead of in registry.Store, so putID can be
+	// declared at the package scope. Otherwise, and anonymous function, or
+	// method, is needed and that will make an allocation.
+	runtime.SetFinalizer(id, putID)
+	return Set{id: id}
+}
+
+var idPool = sync.Pool{New: func() any { return new(uint64) }}
```
Comment (tigrannajaryan) [inline attribute/set.go:85]: Can you clarify why we need a pool of ids? Can we always allocate instead?

File: attribute/set_test.go [hunk 4]
```diff
@@ -225,3 +226,32 @@ func args(m reflect.Method) []reflect.Value {
...
+		attribute.Float64("float64 10", 10),
+		attribute.Float64("float64 -10", -10),
+		attribute.String("string empty", ""),
+		attribute.String("hello", "world"),
+		attribute.BoolSlice("[]bool", []bool{true, false, true}),
+		attribute.Int64Slice("[]int64", []int64{-10, 13124, -23, 0, 2}),
+		attribute.Float64Slice("[]float64", []float64{10.23, 941.1, 184e9, -2.3}),
+		attribute.StringSlice("[]string", []string{"", "one", "two"}),
+	}
+	// Pre-sort to remove from first iteration results.
```
Comment (tigrannajaryan) [inline attribute/set_test.go:247]: Doesn't this defeat the purpose of the benchmark? Aren't we interested in finding out the performance of a typical case where the attributes are not necessarily sorted? Or is being sorted more typical?

File: attribute/set.go [hunk 5]
```diff
@@ -53,16 +58,44 @@ type (
...
 	}
+	*v = (*v)[:length]
+	return v
+}
+
+// sets is a registry of all Set data.
+var sets = newSetRegistry(-1)
+
+func newSet(data *[]KeyValue) Set {
+	id := getID()
```
Comment (tigrannajaryan) [inline attribute/set.go:76]: I am not quite sure but is it necessary to allocate (or get from a pool) a pointer to an id? The id is already stored in the `setData.key`. Can Set.id point to `setData.key` instead?

File: attribute/set.go [hunk 6]
```diff
@@ -29,14 +29,19 @@ type (
 	// This type supports the Equivalent method of comparison using values of
 	// type Distinct.
 	Set struct {
-		equivalent Distinct
+		// id is the key to the sets registry where this set's data is.
+		//
+		// A pointer is used so the finalizer can handle reference-counting for
+		// the sets registry while still being optimized as a map key.
+		id *uint64
```
Comment (tigrannajaryan) [inline attribute/set.go:36]: Would it be possible to make this work by using a pointer to `setData` instead of a pointer to a separate `uint64`?

**Review by tigrannajaryan — COMMENTED** [review]
Sorry, I only had a chance to do a cursory look, so mostly superficial comments.

**Comment by MrAlias** [comment]
The use of finalizers here does not seem like a production ready solution. Closing.

**Pull request closed without merging**